src/protego.py

```diff
@@ -0,0 +1,248 @@
+import re
+from datetime import time
+from collections import namedtuple
+from six.moves.urllib.parse import urlparse, urlunparse, ParseResult, quote, unquote
+
+Record = namedtuple('Record', ['field', 'value'])
+RequestRate = namedtuple(
+    'RequestRate', ['requests', 'seconds', 'start_time', 'end_time'])
+
+DISALLOW_DIRECTIVE = ['disallow', 'dissallow', 'dissalow', 'disalow', 'diasllow', 'disallaw']
+ALLOW_DIRECTIVE = ['allow']
+USER_AGENT_DIRECTIVE = ['user-agent', 'useragent', 'user agent']
+SITEMAP_DIRECTIVE = ['sitemap', 'sitemaps', 'site-map']
+CRAWL_DELAY_DIRECTIVE = ['crawl-delay', 'crawl delay']
+REQUEST_RATE_DIRECTIVE = ['request-rate', 'request rate']
+
+
+def is_valid_directive(d):
+    return d in DISALLOW_DIRECTIVE or d in ALLOW_DIRECTIVE \
+        or d in USER_AGENT_DIRECTIVE or d in SITEMAP_DIRECTIVE \
+        or d in CRAWL_DELAY_DIRECTIVE or d in REQUEST_RATE_DIRECTIVE
+
+
+class RecordGroup:
+
+    def __init__(self):
+        self.user_agents = []
+        self.rules = []
+        self.crawl_delay = None
+        self.sorted_rules = None
+        self.req_rate = None
+
+    def add_user_agent(self, user_agent):
+        user_agent = user_agent.strip().lower()
+        if user_agent != '*' and '*' in user_agent:
+            user_agent = user_agent.replace('*', '')
+        self.user_agents.append(user_agent)
+
+    def applies_to(self, robotname):
```

Comment:
With the current implementation, it looks to me like the returned value may not be the longest match, since we are returning the match length as soon as there is a match, and `self.user_agents` does not seem ordered by length.

Reply:
I have rewritten a good portion of the parser since it was necessary to achieve compatibility with google's parser. This issue is resolved now.